### PR TITLE
Add record component completion. Fix #667

### LINKED ISSUE #667: No completions for records header first components
```
public record Person(|) {}
```

At the position | there are no type completion suggestions are made. Neither after typing a Type like

```
public record Person(String n|) {}
```

No variable name completions or any.

But things start to work after completing the first component

```
public record Person(String name, |) {}
```

Eclipse version: 2022-12
Project Runtime: Java 17

### PR DESCRIPTION
**Pull request by gayanper**
## What it does
Fix record component type and name completions by recovering it as a local field on the recovered type. This follows the strategy of how fields completions are handled by CompletionParser.

## How to test
Completions on record component as described in https://github.com/eclipse-jdt/eclipse.jdt.core/issues/667
## Author checklist

- [x] I have thoroughly tested my changes
- [x] The change is following the [coding conventions](https://wiki.eclipse.org/Platform/How_to_Contribute#Coding_Conventions)
- [x] I have signed the [Eclipse Contributor Agreement (ECA)](https://www.eclipse.org/legal/ECA.php)

### CODE REVIEW

File: org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/parser/Parser.java [hunk 1]
```diff
@@ -93,7 +93,7 @@ public class Parser implements TerminalTokens, ParserBasicInformation, Conflicte
 
 	public static short check_table[] = null;
 	public static final int CurlyBracket = 2;
-	private static final boolean DEBUG = false;
+	private static final boolean DEBUG = true;
```
[the remark this answers is not in the document]
Comment (gayanper) [reply org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/parser/Parser.java:96]: Yes

**Comment by gayanper** [comment]
@stephan-herrmann @iloveeclipse please have a look at the changes.

File: org.eclipse.jdt.core.tests.model/src/org/eclipse/jdt/core/tests/model/CompletionTests14.java [hunk 2]
```diff
@@ -23,7 +23,7 @@
 public class CompletionTests14 extends AbstractJavaModelCompletionTests {
 
 	static {
-		//		TESTS_NAMES = new String[]{"test034"};
+		// TESTS_NAMES = new String[]{"test034"};
```
Comment (stephan-herrmann) [inline org.eclipse.jdt.core.tests.model/src/org/eclipse/jdt/core/tests/model/CompletionTests14.java:26]: ? :)

File: org.eclipse.jdt.core.tests.model/src/org/eclipse/jdt/core/tests/model/CompletionTests14.java [hunk 3]
```diff
@@ -518,4 +518,61 @@ public void testBug553097_5() throws JavaModelException {
...
+				"/Completion/src/Name.java",
+				"public class Name {\n"
+						+ "}\n");
+		CompletionTestsRequestor2 requestor = new CompletionTestsRequestor2(true, true, true, false);
+		String str = this.workingCopies[0].getSource();
+		String completeBehind = "Name ";
+		int cursorLocation = str.lastIndexOf(completeBehind) + completeBehind.length();
+		this.workingCopies[0].codeComplete(cursorLocation, requestor, this.wcOwner);
+		assertResults(
+				"name[VARIABLE_DECLARATION]{name, null, LName;, null, null, name, null, [26, 26], 48}",
```
Comment (stephan-herrmann) [inline org.eclipse.jdt.core.tests.model/src/org/eclipse/jdt/core/tests/model/CompletionTests14.java:574]: I see that some other tests in this class look similar, but wouldn't it be interesting to actually document in the test, how the `relevance` adds up from individual constants? Cf., e.g., `testBug553097_?()`

**Review by stephan-herrmann — COMMENTED** [review]
Good job, as usual :)

You might persuade me to defer additional details to another PR, but perhaps we can go the extra mile just now?

File: org.eclipse.jdt.core/codeassist/org/eclipse/jdt/internal/codeassist/complete/CompletionOnRecordComponentName.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+ *     Gayan Perera - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.jdt.internal.codeassist.complete;
+
+import org.eclipse.jdt.core.compiler.CharOperation;
+import org.eclipse.jdt.internal.compiler.ast.RecordComponent;
+import org.eclipse.jdt.internal.compiler.ast.TypeReference;
+import org.eclipse.jdt.internal.compiler.lookup.BlockScope;
+
+// copied implementation from org.eclipse.jdt.core/codeassist/org/eclipse/jdt/internal/codeassist/complete/CompletionOnFieldName.java
```
[the remark this answers is not in the document]
Comment (gayanper) [reply org.eclipse.jdt.core/codeassist/org/eclipse/jdt/internal/codeassist/complete/CompletionOnRecordComponentName.java:21]: Yes its similar, But if we try to use that, we might not be able to use the record component it self for name completion like we do for field and argument right ? Or did you had something different in mind to handle this ?

File: org.eclipse.jdt.core/codeassist/org/eclipse/jdt/internal/codeassist/complete/CompletionParser.java [hunk 5]
```diff
@@ -415,6 +416,11 @@ protected void attachOrphanCompletionNode(){
 					this.currentElement = this.currentElement.add(fieldDeclaration, 0);
 					return;
 				}
+			} else if (recoveredType.typeDeclaration.isRecord() && orphan instanceof TypeReference) {
+				// translate into a field type completion on the recovered type
+				CompletionOnFieldType fieldType = new CompletionOnFieldType((TypeReference) orphan, true);
+				this.currentElement = this.currentElement.add(fieldType, 0);
+				return;
```
[the remark this answers is not in the document]
Comment (gayanper) [reply org.eclipse.jdt.core/codeassist/org/eclipse/jdt/internal/codeassist/complete/CompletionParser.java]: Agree and Improved it

File: org.eclipse.jdt.core/codeassist/org/eclipse/jdt/internal/codeassist/complete/CompletionParser.java [hunk 6]
```diff
@@ -2247,6 +2253,12 @@ private boolean checkRecoveredType() {
 				this.lastCheckPoint = this.assistNode.sourceEnd + 1;
 				this.isOrphanCompletionNode = true;
 				return true;
+			} else if (this.lastIgnoredToken == -1 && recoveredType.typeDeclaration.isRecord()) {
+				// if we are in a record header
+				this.assistNode = this.getTypeReference(0);
+				this.lastCheckPoint = this.assistNode.sourceEnd + 1;
+				this.isOrphanCompletionNode = true;
+				return true;
```
[the remark this answers is not in the document]
Comment (gayanper) [reply org.eclipse.jdt.core/codeassist/org/eclipse/jdt/internal/codeassist/complete/CompletionParser.java]: Good point, Improved it with the same strategy as above

**Comment by gayanper** [comment]
@stephan-herrmann as always valuable comments, I improved the code a bit more according to suggestion which leads to removing much duplicates.

File: org.eclipse.jdt.core/codeassist/org/eclipse/jdt/internal/codeassist/complete/CompletionOnRecordComponentName.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (stephan-herrmann) [reply org.eclipse.jdt.core/codeassist/org/eclipse/jdt/internal/codeassist/complete/CompletionOnRecordComponentName.java:21]: My comment was just about your comment ("copied ..."). Nothing to change in the code...

**Review by stephan-herrmann — APPROVED** [review]
Thanks for improvements. 

I thought I had commented one more location, but that seems to have gotten lost in github, nothing serious, though.

I'm just surprised about your continued force pushes. Didn't we decide against that in "non-private" branches?

Anyway: :+1:

**Comment by gayanper** [comment]
> Thanks for improvements.
> 
> I thought I had commented one more location, but that seems to have gotten lost in github, nothing serious, though.
> 
> I'm just surprised about your continued force pushes. Didn't we decide against that in "non-private" branches?
> 
> Anyway: 👍

Agree, I will keep commits going forward, when i look back about that discussion, I have also suggested not to force push, seems like i have a bad habit which needs to be corrected.

**Pull request merged**